agent/submatview/store.go

```diff
@@ -0,0 +1,222 @@
+package submatview
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+
+	"github.com/hashicorp/consul/agent/cache"
+	"github.com/hashicorp/consul/lib/ttlcache"
+)
+
+type Store struct {
+	logger     hclog.Logger
+	lock       sync.RWMutex
+	byKey      map[string]entry
+	expiryHeap *ttlcache.ExpiryHeap
+
+	// idleTTL is the duration of time an entry should remain in the Store after the
+	// last request for that entry has been terminated. It is a field on the struct
+	// so that it can be patched in tests without need a lock.
+	idleTTL time.Duration
+}
+
+type entry struct {
+	materializer *Materializer
+	expiry       *ttlcache.Entry
+	stop         func()
+	// requests is the count of active requests using this entry. This entry will
+	// remain in the store as long as this count remains > 0.
+	requests int
+}
+
+func NewStore(logger hclog.Logger) *Store {
+	return &Store{
+		logger:     logger,
+		byKey:      make(map[string]entry),
+		expiryHeap: ttlcache.NewExpiryHeap(),
+		idleTTL:    20 * time.Minute,
+	}
+}
+
+// Run the expiration loop until the context is cancelled.
+func (s *Store) Run(ctx context.Context) {
+	for {
+		s.lock.RLock()
+		timer := s.expiryHeap.Next()
+		s.lock.RUnlock()
+
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+			return
+		case <-s.expiryHeap.NotifyCh:
+			timer.Stop()
+			continue
```

Comment:
Why is this case branch here, and why does it just resume the `for` loop? Is it because this is a notification that the timer on the heap has been updated and we need to recycle the timer?

I think these branches could benefit from some light commentary while it's fresh in your mind.

Reply:
Right, exactly. The heap has changed so we need to start a new timer. I've added comments to these branches.